tests/unit/test_spaces.cpp

```diff
@@ -0,0 +1,122 @@
+#include "gtest/gtest.h"
+#include "VecSim/spaces/space_aux.h"
+#include "VecSim/spaces/IP/IP.h"
+#include "VecSim/spaces/IP/IP_SSE.h"
+#include "VecSim/spaces/IP/IP_AVX.h"
+#include "VecSim/spaces/IP/IP_AVX512.h"
+#include "VecSim/spaces/L2/L2.h"
+#include "VecSim/spaces/L2/L2_SSE.h"
+#include "VecSim/spaces/L2/L2_AVX.h"
+#include "VecSim/spaces/L2/L2_AVX512.h"
+
+class SpacesTest : public ::testing::Test {
+
+protected:
+    SpacesTest() {}
+
+    ~SpacesTest() override {}
+
+    void SetUp() override {}
+
+    void TearDown() override {}
+};
+
+// This test will trigger the "Residuals" function for dimension > 16, for each optimization.
+TEST_F(SpacesTest, l2_17) {
+    Arch_Optimization optimization = getArchitectureOptimization();
+    size_t dim = 17;
+    float v[dim];
+    for (size_t i = 0; i < dim; i++) {
+        v[i] = (float)i;
+    }
+
+    float baseline = L2Sqr(v, v, &dim);
+    switch (optimization) {
+    case ARCH_OPT_AVX512:
+        ASSERT_EQ(baseline, L2SqrSIMD16ExtResiduals_AVX512(v, v, &dim));
+        optimization = ARCH_OPT_AVX;
+    case ARCH_OPT_AVX:
+        ASSERT_EQ(baseline, L2SqrSIMD16ExtResiduals_AVX(v, v, &dim));
+        optimization = ARCH_OPT_SSE;
+    case ARCH_OPT_SSE:
+        ASSERT_EQ(baseline, L2SqrSIMD16ExtResiduals_SSE(v, v, &dim));
+        break;
+    default:
+        ASSERT_TRUE(false);
+    }
+}
+
+// This test will trigger the "Residuals" function for dimension < 9, for each optimization.
+TEST_F(SpacesTest, l2_9) {
+    Arch_Optimization optimization = getArchitectureOptimization();
+    size_t dim = 9;
+    float v[dim];
+    for (size_t i = 0; i < dim; i++) {
+        v[i] = (float)i;
+    }
+
+    float baseline = L2Sqr(v, v, &dim);
+    switch (optimization) {
+    case ARCH_OPT_AVX512:
+        ASSERT_EQ(baseline, L2SqrSIMD16ExtResiduals_AVX512(v, v, &dim));
+        optimization = ARCH_OPT_AVX;
+    case ARCH_OPT_AVX:
+        ASSERT_EQ(baseline, L2SqrSIMD16ExtResiduals_AVX(v, v, &dim));
+        optimization = ARCH_OPT_SSE;
+    case ARCH_OPT_SSE:
+        ASSERT_EQ(baseline, L2SqrSIMD4ExtResiduals_SSE(v, v, &dim));
+        break;
```

Comment:
What is the difference from the previous test?

Reply:
the vector length compared to 16 and 4. This is how to space choses the best function